cloudigrade/account/util.py

```diff
@@ -126,80 +135,134 @@ def generate_aws_ami_messages(instances_data, ami_list):
-        producer = conn.Producer(serializer='json')
-        for message in messages:
-            producer.publish(
-                message,
-                retry=True,
-                exchange=exchange,
-                routing_key=queue_name,
-                declare=[message_queue]
-            )
+    wrapped_messages = [_sqs_wrap_message(message) for message in messages]
+    batch_count = math.ceil(len(messages) / SQS_SEND_BATCH_SIZE)
+
+    for batch_num in range(batch_count):
+        start_pos = batch_num * SQS_SEND_BATCH_SIZE
+        end_pos = start_pos + SQS_SEND_BATCH_SIZE - 1
+        batch = wrapped_messages[start_pos:end_pos]
+        sqs.send_message_batch(QueueUrl=queue_url, Entries=batch)
 
 
 def read_messages_from_queue(queue_name, max_count=1):
     """
-    Read messages (up to max_count) from a message queue.
+    Read messages (up to max_count) from an SQS queue.
 
     Args:
         queue_name (str): The queue to read messages from
         max_count (int): Max number of messages to read
 
     Returns:
-        list[object]: The dequeued messages.
+        list[object]: The de-queued messages.
 
     """
-    __, message_queue = _create_exchange_and_queue(queue_name)
+    queue_url = _get_sqs_queue_url(queue_name)
+    sqs = boto3.client('sqs')
+    sqs_messages = []
+    max_batch_size = min(SQS_RECEIVE_BATCH_SIZE, max_count)
+    for __ in range(max_count):
+        # Because receive_message does *not* actually reliably return
+        # MaxNumberOfMessages number of messages especially (read the docs),
+        # our iteration count is actually max_count and we have some
+        # conditions at the end that break us out when we reach the true end.
+        new_messages = sqs.receive_message(
+            QueueUrl=queue_url,
+            MaxNumberOfMessages=max_batch_size
+        ).get('Messages', [])
+        if len(new_messages) == 0:
+            break
+        sqs_messages.extend(new_messages)
+        if len(sqs_messages) >= max_count:
+            break
     messages = []
-    with kombu.Connection(settings.CELERY_BROKER_URL) as conn:
+    for sqs_message in sqs_messages:
         try:
-            consumer = conn.SimpleQueue(name=message_queue)
-            while len(messages) < max_count:
-                message = consumer.get_nowait()
-                messages.append(message.payload)
-                message.ack()
-        except Empty:
-            pass
+            unwrapped = _sqs_unwrap_message(sqs_message)
+            sqs.delete_message(
+                QueueUrl=queue_url,
+                ReceiptHandle=sqs_message['ReceiptHandle'],
+            )
+            messages.append(unwrapped)
+        except ClientError as e:
+            # I'm not sure exactly what exceptions could land here, but we
+            # probably should log them, stop attempting further deletes, and
+            # return what we have received (and thus deleted!) so far.
+            logger.exception(e)
```

Comment:
I think it would be helpful to add context to the error log.  What queue had failures?

Reply:
Good idea! I'm adding more error logging in the next commit.